hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/HapiSpec.java

```diff
@@ -457,6 +485,132 @@ private void exec(List<HapiSpecOperation> ops) {
+        final var scheduleStateAssertion = getScheduleInfo(schedule);
+        // If the original transaction was supposed to fail with INVALID_SIGNATURE,
+        // then the schedule should not have been executed
+        if (expectedStatus != INVALID_SIGNATURE) {
+            scheduleStateAssertion.isExecuted();
+        } else {
+            scheduleStateAssertion.isNotExecuted();
+        }
+        orderedOps[orderedOps.length - 2] = scheduleStateAssertion;
+
+        if (expectedStatus != INVALID_SIGNATURE) {
+            final var recordAssertion = getTxnRecord(creation)
+                    .scheduledBy(schedule)
+                    .hasPriority(recordWith().status(expectedStatus));
+            orderedOps[orderedOps.length - 1] = recordAssertion;
+        } else {
+            // If the original transaction was supposed to fail with INVALID_SIGNATURE, there
+            // will be no scheduled transaction record to retrieve, so just use noOp()
+            orderedOps[orderedOps.length - 1] = noOp();
+        }
+        return blockingOrder(orderedOps);
+    }
+
+    private boolean isEmpty(final Key key) {
+        if (key.hasKeyList()) {
+            return key.getKeyList().getKeysCount() == 0
+                    || key.getKeyList().getKeysList().stream().allMatch(this::isEmpty);
+        } else if (key.hasThresholdKey()) {
+            return key.getThresholdKey().getKeys().getKeysCount() == 0
+                    || key.getThresholdKey().getKeys().getKeysList().stream().allMatch(this::isEmpty);
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Given the number of keys that will be used to sign the transaction, and the number of
+     * {@code ScheduleSign} transactions that will be executed, create a list of indices that
+     * can be used to choose which keys to sign with in both the initial {@code ScheduleCreate},
+     * and any {@code ScheduleSign} transactions.
+     *
+     * @param numKeys the number of keys that will be used to sign the transaction
+     * @param numSignTxns the number of {@code ScheduleSign} transactions that will be executed
+     * @return a list of indices that can be used to choose signing keys
+     */
+    private static List<Integer> createAndSignIndicesGiven(final int numKeys, final int numSignTxns) {
+        final List<Integer> endIndices = new ArrayList<>();
+        endIndices.add(numKeys);
+        int remainingSkipsAllowed = numKeys - numSignTxns;
+        for (int i = 0; i < numSignTxns; i++) {
+            final var skips = remainingSkipsAllowed > 0 ? RANDOM.nextInt(remainingSkipsAllowed) : 0;
+            remainingSkipsAllowed -= skips;
+            final var curIndex = endIndices.get(0);
+            final var nextEndIndex = curIndex - skips - 1;
+            endIndices.add(0, nextEndIndex);
+        }
+        endIndices.add(0, 0);
+        return endIndices;
+    }
```

Comment:
Using a `List<Integer>` to define possible indices for multiple keys is super confusing to me, but then again, so are most things